lib/src/cluster.dart

```diff
@@ -58,46 +58,114 @@ class Cluster implements Client {
       pageSize: pageSize,
       pagingState: pagingState,
     );
-    return _withConnection((c) => c._protocol.query(this, q, body));
+    final peer = _selectPeer();
+    return peer._sendQuery(this, q, body);
   }
 
   Future _connect(String hostPort) async {
-    final c = await _Connection.open(
+    final c = await _Peer.connectAdress(
       hostPort,
       authenticator: _authenticator,
     );
-    _connections.add(c);
+    _peers.add(c);
   }
 
-  Future<R> _withConnection<R>(Future<R> body(_Connection c)) async {
-    _connectionCursor++;
-    if (_connectionCursor >= _connections.length) {
-      _connectionCursor = 0;
+  _Peer _selectPeer() {
+    final latencies = Map.fromIterables(_peers.map((p) => p.latency), _peers);
+    final latenciesSum = latencies.keys.fold<double>(0, (a, b) => a + 1 / b);
+    final rd = Random().nextDouble();
+    double lat = 0;
+    for (final entry in latencies.entries) {
+      lat += 1 / entry.key;
+      if (rd * latenciesSum < lat) {
+        return entry.value;
+      }
+    }
+    // if above fails becase some unknown flotng point error:
+    return _peers[Random().nextInt(_peers.length)];
+  }
+
+  _collectPeers() async {
+    final page = await query('SELECT peer FROM system.peers');
+    final newIPs =
+        page.items.map((row) => row.values[0] as InternetAddress).toList();
+    for (final peer in _peers) {
+      if (newIPs.contains(peer.host)) {
+        newIPs.remove(peer.host);
+      }
+    }
+    for (final ip in newIPs) {
+      final newPeer =
+          await _Peer.connect(ip, 9042, authenticator: _authenticator);
```

Comment:
Also collect the port from the `system.peers`.

Reply:
Cassandra does not show the port in the `system.peers` table, so we have to use the default port.